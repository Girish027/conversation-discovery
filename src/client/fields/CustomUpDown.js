import React, { Component } from 'react';
import PropTypes from 'prop-types';

import ObjectUtils from '../utils/ObjectUtils';
import { customUpDown } from '../styles/customComponentsStyles';
import ValidationPlaceholder from './ValidationPlaceholder';
import GenericTooltip from '../components/Tooltip/Tooltip';
import Language from '../Language';

export default class CustomUpDown extends Component {
  constructor(props) {
    super(props);
    this.schema = props.schema;
    const value = ObjectUtils.isEmptyOrNull(props.formData)
      ? this.schema.default
      : props.formData;
    this.state = {
      value,
      validationMessage: ''
    };
    this.handleChange = this.handleChange.bind(this);
  }

  getLabel() {
    return (
      <div
        style={customUpDown.label}
      >
        {this.schema.title}
      </div>
    );
  }

  getField() {
    const { validationMessage } = this.state;
    let validationMsg = '';

    if (validationMessage) {
      validationMsg = validationMessage.slice(0, validationMessage.indexOf('('));
    }

    return (
      <React.Fragment>
        <GenericTooltip content={this.props.name === 'numOfClusters' ? Language.RUN.TOOLTIP.custers : Language.RUN.TOOLTIP.turns}>
          <input
            value={this.state.value}
            name={this.props.name}
            type='number'
            onChange={this.handleChange}
            min={this.schema.minimum}
            max={this.schema.maximum}
            style={customUpDown.field}
          />
        </GenericTooltip>
        {validationMessage && (
          <ValidationPlaceholder validationMessage={validationMsg} />
        )}
      </React.Fragment>
    );
  }

  handleChange(event) {
    const { value, validationMessage } = event.target;
    const newValue = parseInt(value, 10);
    this.setState({ value: newValue, validationMessage });
    this.props.onChange(newValue);
  }

  render() {
    return (
      <div
        style={customUpDown}
      >
        {this.getLabel()}
        {this.getField()}
      </div>
    );
  }
}

CustomUpDown.propTypes = {
  name: PropTypes.string,
  schema: PropTypes.object.isRequired,
  formData: PropTypes.number,
  onChange: PropTypes.func,
};

CustomUpDown.defaultProps = {
  onChange: () => {},
  name: 'updown',
};
