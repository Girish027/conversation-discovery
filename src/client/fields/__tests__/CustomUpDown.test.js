import React from 'react';
import { shallow } from 'enzyme';
import toJSON from 'enzyme-to-json';
import CustomUpDown from '../CustomUpDown';

describe('<CustomUpDown />', function () {
  let wrapper;

  const props = {
    name: 'test',
    schema: {
      maximum: 900,
      minimum: 2,
      title: 'Clusters (2-900):'
    },
    uiSchema: {}
  };

  afterAll(() => {
    jest.clearAllMocks();
  });

  describe('Creating an instance', () => {
    test('should exist', () => {
      wrapper = shallow(<CustomUpDown
        {...props}
      />);
      expect(wrapper.exists()).toBe(true);
    });
  });

  describe('Snapshots', () => {
    test('renders correctly for default props', () => {
      wrapper = shallow(<CustomUpDown
        {...props}
      />);
      expect(toJSON(wrapper)).toMatchSnapshot();
    });
  });

  describe('Functionality:', () => {
    beforeEach(function () {
      props.onChange = jest.fn();
    });
    describe('onChange', function () {
      test('should pass the props and change the state with new value in the form', () => {
        wrapper = shallow(<CustomUpDown
          {...props}
        />);

        wrapper.instance().handleChange({
          target: {
            name: 'cluster',
            value: '2',
          }
        });
        expect(props.onChange).toHaveBeenCalledWith(2);
        expect(wrapper.state().value).toEqual(2);
      });

      test('should chamge the state in case of validation failure', () => {
        wrapper = shallow(<CustomUpDown
          {...props}
        />);

        expect(wrapper.state().validationMessage).toEqual('');
        wrapper.instance().handleChange({
          target: {
            name: 'cluster',
            value: '1',
            validationMessage: 'Clusters should be greater than 2 char'
          }
        });
        expect(wrapper.state().validationMessage).toEqual('Clusters should be greater than 2 char');
      });
    });
  });
});
