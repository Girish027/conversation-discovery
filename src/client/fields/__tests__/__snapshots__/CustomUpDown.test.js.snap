// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CustomUpDown /> Snapshots renders correctly for default props 1`] = `
<div
  style={
    Object {
      "field": Object {
        "border": "1px solid #DCE1E8",
        "borderRadius": "0 2px 2px 0",
        "display": "block",
        "lineHeight": "25px",
        "paddingLeft": "5px",
        "width": "100%",
      },
      "label": Object {
        "color": "#6E6E6E",
        "fontSize": "14px",
        "fontWeight": "normal",
        "letterSpacing": 0,
        "paddingBottom": "5px",
        "textAlign": "left",
      },
      "marginTop": "20px",
    }
  }
>
  <div
    style={
      Object {
        "color": "#6E6E6E",
        "fontSize": "14px",
        "fontWeight": "normal",
        "letterSpacing": 0,
        "paddingBottom": "5px",
        "textAlign": "left",
      }
    }
  >
    Clusters (2-900):
  </div>
  <GenericTooltip
    content="The Number of user-turns to use â€“ User intents are not always obvious in the first user turns. Increase this turn count may improve the odds of finding useful intents in the transcripts. But bringing in too many user turns may add noise to the intent discovery process. We recommend that the number of user-turns stays between 2 and 3."
    direction="bottom"
    type="info"
  >
    <input
      max={900}
      min={2}
      name="test"
      onChange={[Function]}
      style={
        Object {
          "border": "1px solid #DCE1E8",
          "borderRadius": "0 2px 2px 0",
          "display": "block",
          "lineHeight": "25px",
          "paddingLeft": "5px",
          "width": "100%",
        }
      }
      type="number"
    />
  </GenericTooltip>
</div>
`;
